cmd/autonity/genesis_test.go

```diff
@@ -82,14 +165,145 @@ func TestCustomGenesis(t *testing.T) {
+			"--exec", tt.query, "console")
+		autonity.ExpectRegexp(tt.result)
+		autonity.ExpectExit()
+
+		// Then start node again with same genesis, node should perform a normal start up.
+		autonity = runAutonity(t,
+			"--port", "0", "--maxpeers", "0", "--nodiscover", "--nat", "none",
+			"--etherbase", coinbase, "--ipcpath", ipc, "--datadir", datadir, "--genesis", json)
+
+		defer func() {
+			autonity.Interrupt()
+			autonity.ExpectExit()
+		}()
+
+		// Check RPC endpoint service is loaded and RPC service is a available now.
+		waitForEndpoint(t, ipc, 5*time.Second)
+	}
+}
+
+// Test start node with mismatch genesis block and incompatible chain configuration, the node start up should end up,
+func TestStartNodeWithIncompatibleGenesis(t *testing.T) {
+	testStartNodeWithIncompatibleGenesis := func(t *testing.T, defaultGenesis string, newGenesis string) {
+		for _, tt := range customGenesisTests {
+			coinbase := "0x8605cdbbdb6d264aa742e77020dcbc58fcdce182"
+			var ipc string
+			if runtime.GOOS == "windows" {
+				ipc = `\\.\pipe\autonity` + strconv.Itoa(trulyRandInt(100000, 999999))
+			} else {
+				ws := tmpdir(t)
+				defer os.RemoveAll(ws)
+				ipc = filepath.Join(ws, "autonity.ipc")
+			}
+
+			// Create a temporary data directory to use and inspect later
+			datadir := tmpdir(t)
+			defer os.RemoveAll(datadir)
+
+			// Initialize the data directory with the default genesis block
+			json := filepath.Join(datadir, "genesis.json")
+			if err := ioutil.WriteFile(json, []byte(defaultGenesis), 0600); err != nil {
+				t.Fatalf("failed to write genesis file: %v", err)
+			}
+
+			// First create chain data with genesis, and then stop the node.
+			autonity := runAutonity(t,
+				"--port", "0", "--maxpeers", "0", "--nodiscover", "--nat", "none",
+				"--etherbase", coinbase, "--ipcpath", ipc, "--datadir", datadir, "--genesis", json)
+			waitForEndpoint(t, ipc, 5*time.Second)
+			autonity.Interrupt()
+
+			// Make sure genesis block is created.
+			autonity = runAutonity(t, "--nousb",
+				"--datadir", datadir, "--maxpeers", "0", "--port", "0",
+				"--nodiscover", "--nat", "none", "--ipcdisable",
+				"--exec", tt.query, "console")
+			autonity.ExpectRegexp(tt.result)
+			autonity.ExpectExit()
+
+			// Then start node again with a incompatible genesis file, node should not init a new genesis, and the start up should be interrupted.
+			jsonV2 := filepath.Join(datadir, "genesisV2.json")
+			if err := ioutil.WriteFile(jsonV2, []byte(newGenesis), 0600); err != nil {
+				t.Fatalf("failed to write genesis file: %v", err)
+			}
+			autonity = runAutonity(t,
+				"--port", "0", "--maxpeers", "0", "--nodiscover", "--nat", "none",
+				"--etherbase", coinbase, "--ipcpath", ipc, "--datadir", datadir, "--genesis", jsonV2)
+
+			defer func() {
+				autonity.Interrupt()
+				autonity.ExpectExit()
+			}()
+
+			// With a incompatible genesis, client should do nothing with initGenesis by exit the start up.
+			autonity.ExpectExit()
```

Comment:
The reference to autonity will change at line 296 by the time this defer function is executed and this piece of code would be redundant since on line 301 and 302 will cause the autonity to exit. The autonity reference should be renamed so that the correct autonity is closed.

```suggestion
			autonity1 = runAutonity(t,
				"--port", "0", "--maxpeers", "0", "--nodiscover", "--nat", "none",
				"--etherbase", coinbase, "--ipcpath", ipc, "--datadir", datadir, "--genesis", jsonV2)

			defer func() {
				autonity1.Interrupt()
				autonity1.ExpectExit()
			}()

			// With a incompatible genesis, client should do nothing with initGenesis by exit the start up.
			autonity1.ExpectExit()
```

Reply:
> The reference to autonity will change at line 296 by the time this defer function is executed and this piece of code would be redundant since on line 301 and 302 will cause the autonity to exit. The autonity reference should be renamed so that the correct autonity is closed.

Thank you, Anneeque. It's resolved at my latest commit: Make testcase of initGenesis to be lean.